dev/benchmarks/test_apps/stocks/lib/main.dart

```diff
@@ -65,49 +45,161 @@ class StocksAppState extends State<StocksApp> {
-      );
-    }
-    // The other paths we support are in the routes table.
+    assert(configuration.stockMode != null);
     return null;
   }
 
   @override
   Widget build(BuildContext context) {
+    final StockConfiguration configuration = StockConfigurationProvider.of(context);
     assert(() {
-      debugPaintSizeEnabled = _configuration.debugShowSizes;
-      debugPaintBaselinesEnabled = _configuration.debugShowBaselines;
-      debugPaintLayerBordersEnabled = _configuration.debugShowLayers;
-      debugPaintPointersEnabled = _configuration.debugShowPointers;
-      debugRepaintRainbowEnabled = _configuration.debugShowRainbow;
+      debugPaintSizeEnabled = configuration.debugShowSizes;
+      debugPaintBaselinesEnabled = configuration.debugShowBaselines;
+      debugPaintLayerBordersEnabled = configuration.debugShowLayers;
+      debugPaintPointersEnabled = configuration.debugShowPointers;
+      debugRepaintRainbowEnabled = configuration.debugShowRainbow;
       return true;
     }());
-    return MaterialApp(
+    return MaterialApp.router(
       title: 'Stocks',
-      theme: theme,
+      theme: _getThemeFromConfiguration(configuration),
       localizationsDelegates: StockStrings.localizationsDelegates,
       supportedLocales: StockStrings.supportedLocales,
-      debugShowMaterialGrid: _configuration.debugShowGrid,
-      showPerformanceOverlay: _configuration.showPerformanceOverlay,
-      showSemanticsDebugger: _configuration.showSemanticsDebugger,
-      routes: <String, WidgetBuilder>{
-         '/':         (BuildContext context) => StockHome(stocks, _configuration, configurationUpdater),
-         '/settings': (BuildContext context) => StockSettings(_configuration, configurationUpdater),
-      },
-      onGenerateRoute: _getRoute,
+      debugShowMaterialGrid: configuration.debugShowGrid,
+      showPerformanceOverlay: configuration.showPerformanceOverlay,
+      showSemanticsDebugger: configuration.showSemanticsDebugger,
+      routeNameParser: StockRouteNameParser(),
+      routerDelegate: StockRouterDelegate(_state),
+    );
+  }
+}
+
+class StockRouteNameParser extends RouteNameParser<Uri> {
+  @override
+  Future<Uri> parse(String routeName) {
+    final List<String> segments = routeName.split('?');
+    final String path = segments[0];
+    final String query = segments.length > 1 ? segments[1] : '';
+    final Uri syntheticUrl = Uri(path: path, query: query);
+    return SynchronousFuture<Uri>(syntheticUrl);
+  }
+
+  @override
+  String restore(Uri configuration) {
+    return configuration.toString();
+  }
+}
+
+class StockRouterDelegate extends RouterDelegate<Uri> with PopNavigatorRouterDelegateMixin<Uri>{
+  StockRouterDelegate(
+    this.stockState
+  ) : navigatorKey = GlobalObjectKey<NavigatorState>(stockState) {
+    stockState.addListener(notifyListeners);
+  }
+  StockState stockState;
+
+  final HeroController _heroController = MaterialApp.createMaterialHeroController();
```

Comment:
Is there some way we can avoid having the developers have to do this? It certainly is a lot less magical if the hero controller is not automatically added by MaterialApp.

Maybe allowing navigators to be provided by inherited widget? (as a separate PR that lands before this one)

Though even better would be to find a way to just get rid of observers entirely...

Reply:
We talked about this, one of the option back them is hosting navigator observer and making the navigator retrieve them directly. However, that will be messy to clean up if we eventually want to make the navigator observer goes away. Refactoring the Hero controller to not use navigator observer is a rather big project, and we were thinking we should not touch that in the navigator 2.0

I don't think there is a clean intermediate step we can take.

Reply:
thinking about this over the last few days, the way i see it, we need to notify something in a tree about something else in the tree. The messages either have to go up the tree or down the tree or direct to the target. Up the tree means we bubble messages with the notification system (either from the heroes to the navigator or from the navigator to the MaterialApp or some such). Down the tree means we have a notifier of some sort, using an inherited widget of some sort, either from the MaterialApp to the Navigator or the Navigator to the Heroes. Direct to the target means we provide the navigator with a link to the MaterialApp (what we do now).

Given that we're losing the close ties between the MaterialApp and the navigator, it seems hard to cleanly do the direct link anymore. This suggests we should use one of the other ways.

I don't think talking to the Heroes really works since we need a coordinator to actually do the work of coordinating the Heroes.

So that leaves bubbling messages up from the Navigator to the MaterialApp or sending messages down via inherited widgets from the MaterialApp to the Navigator.

cc @goderbauer @HansMuller

Reply:
For bubbling up the message, do you mean using the same mechanism as scroll notification?

We can still do direct link by completely separate out the hero from navigator if it no longer rely on route notification. Instead of actively listens to route change in navigator, the herocontroller will be passively waiting for hero widget to register flight manifest with. The hero widget will be in charge to observe the route change either via RouteAware or implementing a hook from modal route